Add did:psqr method.

gribneau opened the pull request:
----- DID METHOD REGISTRATION FORM: DELETE EVERYTHING ABOVE THIS LINE ------

### DID Method Registration

As a DID method registrant, I have ensured that my DID method registration complies with the following statements:

- [y] The DID Method specification [defines the DID Method Syntax](https://w3c.github.io/did-core/#method-syntax).
- [y] The DID Method specification [defines the Create, Read, Update, and Deactivate DID Method Operations](https://w3c.github.io/did-core/#method-operations).
- [y] The DID Method specification [contains a Security Considerations section](https://w3c.github.io/did-core/#security-requirements).
- [y] The DID Method specification [contains a Privacy Considerations section](https://w3c.github.io/did-core/#privacy-requirements).
- [y] The JSON file I am submitting has [passed all automated validation tests below](#partial-pull-merging).
- [y] The JSON file contains a `contactEmail` address [OPTIONAL].
- [y] The JSON file contains a `verifiableDataRegistry` entry [OPTIONAL].

OR13 reviewed (commented):
Looks awesome, its very similar to did:web and did:onion.

I think the "verifiableDataRegistry" is "DNS + Websites", but the field is optional, so the feedback is non blocking.

You might consider adding some DID URL examples to make the use of the `/` in the method specific identifier more explicit, for example, see: https://www.w3.org/TR/did-core/#path

it's not obvious to me how you would distinguish between DID URLs with paths and DIDs that have a path component.

```
idchar             = ALPHA / DIGIT / "." / "-" / "_" / pct-encoded
```

To me this implies your examples violate the ABNF, since they make use of `/`.

Assuming the ABNF for the Method does not violate DID Core, I am in support of merging, but I want to double check with folks who have more experience with this part of the spec.

cc @peacekeeper @dlongley @talltree  for review.

I will wait for reviews before approving.

gribneau commented:
>  I think the "verifiableDataRegistry" is "DNS + Websites"
If I tie it back, I'll probably be a bit more explicit about the SSL layer.

> it's not obvious to me how you would distinguish between DID URLs with paths and DIDs that have a path component.
DID path semantics are unclear to me. I could be more explicit that this method uses the DID path as the URL path. That is explicitly supported, but I did not call it out with clarity.

gribneau commented:
Upon reflection, this avoids confusion:
![image](https://user-images.githubusercontent.com/250138/149600060-32508db9-2f3e-4ffe-92cf-9e201d2b8e3a.png)

peacekeeper commented:
Usually we ask for "Security Considerations" and "Privacy Considerations" to be separate sections, maybe you could split that up?

Otherwise looks good I think..

gribneau commented:
> Usually we ask for "Security Considerations" and "Privacy Considerations" to be separate sections, maybe you could split that up?

Done.

peacekeeper commented:
Is this essentially what did:web would have become if https://github.com/w3c-ccg/did-method-web/pull/47 had been merged?

Nothing wrong with this, just curious!

iherman commented:
> Upon reflection, this avoids confusion:
> ![image](https://user-images.githubusercontent.com/250138/149600060-32508db9-2f3e-4ffe-92cf-9e201d2b8e3a.png)

I just wonder whether it is possible to avoid reusing the same term at least in the specification. The risk of confusion with the `path` feature in the DID spec is big, and we should better avoid it imho.

gribneau commented:
> I just wonder whether it is possible to avoid reusing the same term at least in the specification.

The DID core specification treats the path element in precisely the same fashion as generic URIs, with one critical exception.

The explicit similarity is laid out in these sections:
https://www.w3.org/TR/did-core/#path
https://www.w3.org/TR/did-core/#dfn-did-paths

This phrase summarizes nicely:
> A DID path is identical to a generic URI path...

While DID paths may be identical to generic URI paths in appearance, they differ in function. A DID path identifies a secondary resource by reference to the primary resource while a URI path identifies the primary resource. The DID path looks like a URI path, but behaves like a URI fragment.

This becomes apparent if one compares these RFC sections:
https://www.rfc-editor.org/rfc/rfc3986#section-3.3
https://www.rfc-editor.org/rfc/rfc3986#section-3.5

in view of this language in the DID core specification: 
![image](https://user-images.githubusercontent.com/250138/150335802-2752d5ed-9b1a-4c35-854b-7d793081c796.png)

I agree that this will certainly cause confusion. I initially misread the core specification on path myself as a result of the divergence.

The two resolutions that make sense to me would be revising the treatment of path in the DID core specification to be consistent with generic URI paths, or to use a different term.

For the moment, I've revised the language of did:psqr to explicitly reference URL path as a component of the HTTPS URL.

gribneau commented:
> Is this essentially what did:web would have become if [w3c-ccg/did-method-web#47](https://github.com/w3c-ccg/did-method-web/pull/47) had been merged?
> 
> Nothing wrong with this, just curious!

I would have preferred to use did:web, but I wasn't willing to sacrifice standard web architecture and the elegant integration it affords with existing web infrastructure in the process.

iherman commented:
> The two resolutions that make sense to me would be revising the treatment of path in the DID core specification to be consistent with generic URI paths…

There was a long discussion in the WG back then about the differentiation between a DID and a DID URL (which the path is part of). I really do not think we should reopen this...

> …or to use a different term

I see that the specifications says (as opposed from the quote in https://github.com/w3c/did-spec-registries/pull/406#issuecomment-1013552503):

> `psqr-did = "did:psqr:" domain-name`
> `psqr-did = "did:psqr:" domain-name ? (":" url-path)`

Hopefully that (plus the remark later in that section) will be enough to avoid confusion. Thanks.

OR13 commented:
Here is a potential example I made from replacing the did in a did key document with the did, from your spec:

```json
{
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "https://w3id.org/security/suites/jws-2020/v1"
    ],
    "id": "did:psqr:example.com:user",
    "verificationMethod": [
      {
        "id": "did:psqr:example.com:user#z6MknuE5gD3VNy1vkJsQynZzDbPGRKturoxiMeXpMLTw6iWU",
        "type": "JsonWebKey2020",
        "controller": "did:psqr:example.com:user",
        "publicKeyJwk": {
          "kty": "OKP",
          "crv": "Ed25519",
          "x": "fYRgLbA8NxUwu8jm-DguH8U7U1_ycTQms4xToC65v5k"
        }
      },
      {
        "id": "did:psqr:example.com:user#z6LStqsASqk1dcZ2tTckQ6yUfEkf4jSCD1MCSKmuj8SCwNGq",
        "type": "JsonWebKey2020",
        "controller": "did:psqr:example.com:user",
        "publicKeyJwk": {
          "kty": "OKP",
          "crv": "X25519",
          "x": "_xIhtUpp-iQNkWFEZcSSvZFH-2JRpUs0glIzoM9L2Qo"
        }
      }
    ],
    "assertionMethod": [
      "did:psqr:example.com:user#z6MknuE5gD3VNy1vkJsQynZzDbPGRKturoxiMeXpMLTw6iWU"
    ],
    "authentication": [
      "did:psqr:example.com:user#z6MknuE5gD3VNy1vkJsQynZzDbPGRKturoxiMeXpMLTw6iWU"
    ],
    "capabilityInvocation": [
      "did:psqr:example.com:user#z6MknuE5gD3VNy1vkJsQynZzDbPGRKturoxiMeXpMLTw6iWU"
    ],
    "capabilityDelegation": [
      "did:psqr:example.com:user#z6MknuE5gD3VNy1vkJsQynZzDbPGRKturoxiMeXpMLTw6iWU"
    ],
    "keyAgreement": [
      "did:psqr:example.com:user#z6LStqsASqk1dcZ2tTckQ6yUfEkf4jSCD1MCSKmuj8SCwNGq"
    ]
  }
  ```
  
  Note that. `/alice` is not present in the did document.
  
  Per the spec: https://www.w3.org/TR/did-core/#did-subject
  
  > The value of id MUST be a string that conforms to the rules in § 3.1 DID Syntax and MUST exist in the root map of the data model for the DID document.
  
  https://www.w3.org/TR/did-core/#did-syntax (does not include `url-path` or the characters necessary to encode one, afaik).
  
 FWIW I agree with pretty much everything you have said regarding `path` in DID URLs... and I have never seen `path` be used in a DID URL.

gribneau commented:
> @gribneau would you be willing to add an example did document to the spec?

I added the sample DID Document.

pchampin commented:
> Per the spec: https://www.w3.org/TR/did-core/#did-subject
> 
> > The value of id MUST be a string that conforms to the rules in § 3.1 DID Syntax and MUST exist in the root map of the data model for the DID document.
> 
> https://www.w3.org/TR/did-core/#did-syntax (does not include `url-path` or the characters necessary to encode one, afaik).

Yes, I find this part of the PSQR method specification confusing. As I read it, its defined method-specific-id includes `url-path`, which contradicts the syntax of DIDs, as pointed out by @OR13 above. 

More specifically: `did:psqr:example.com:user/alice` is cited as an example of PQSR DID, while it is not a valid DID (although it is a valid DID URL).

gribneau commented:
Yes, @pchampin, paths are a problem in the DID core specification. [As noted above](https://github.com/w3c/did-spec-registries/pull/406#issuecomment-1017437484),  the handling of URI paths and queries in DID core is inconsistent with the underlying RFC.

For the moment, I will remove examples including normal URI paths. I will not, emphatically, add such examples with percent encoding of the `/` path delimiter.

Nobody should interpret this action as agreement on my part that DID paths and queries should behave as URI fragments rather than URI paths or URI queries.

For the record, I find it objectionable.

I also find it odd. Do we really need three URI elements - path, query, and fragment - to identify a section of the same document? I am not surprised that @OR13 has never seen DID path used in this fashion.

msporny reviewed (approved):
Meets minimum requirements for registration.

davidlehn commented:
@gribneau Some notes related to your JSON-LD usage:
- The JSON-LD context at https://vpsqr.com/ns/did-psqr/v1/ needs some work.  It's not valid JSON, fix the commas.  You probably mean "psqr" to be scoped?  Add an `@id` property and put the other properties under an embedded `@context` object.  Defining things as ".../v1/#property" might not be the best thing as far as future updates.
- The example in https://vpsqr.com/did-psqr/v1/ is not valid JSON, fix the commas.  It's also missing a context that defines the terms used.  The processed output semantic data is mostly empty.

kdenhartog commented on methods/psqr.json:
```diff
@@ -0,0 +1,9 @@
+    "status": "registered",
+    "verifiableDataRegistry": "Public Square",
+    "contactName": "Christian Gribneau",
+    "contactEmail": "protocol@vpsqr.com",
```
Are you good with editors assuming that anyone with this email domain is acceptable to make updates to this registration entry?

gribneau replied:
> Are you good with editors assuming that anyone with this email domain is acceptable to make updates to this registration entry?

The email address `protocol@vpsqr.com` owns the open source code, standards, and distributions. It is a role that distributes to myself and one other person in the company.

kdenhartog reviewed (approved):
Status: Approving on the basis that all minimum requirements have been met. 

Additional reviewer notes:

Skipping this since I can tell that the author has a good understanding of the intricate nuances that will pop up with interoperability. @gribneau let me know if you want me to take a finer toothed comb to the method though.

gribneau commented:
> @gribneau Some notes related to your JSON-LD usage:
> 
> * The JSON-LD context at https://vpsqr.com/ns/did-psqr/v1/ needs some work.  It's not valid JSON, fix the commas.  You probably mean "psqr" to be scoped?  Add an `@id` property and put the other properties under an embedded `@context` object.  Defining things as ".../v1/#property" might not be the best thing as far as future updates.
> * The example in https://vpsqr.com/did-psqr/v1/ is not valid JSON, fix the commas.  It's also missing a context that defines the terms used.  The processed output semantic data is mostly empty.

Thanks for the review @davidlehn.

> fix the commas.
Done.

The JSON LD is only human readable at this point. Once it rises to the top of the development queue, we'll flesh it out and make it functional.

I did scope the `@context` under `psqr` and add `@id`, but a functional implementation of JSON LD is not yet present.

davidlehn commented:
> > @gribneau Some notes related to your JSON-LD usage:
> > 
> > * The JSON-LD context at https://vpsqr.com/ns/did-psqr/v1/ needs some work.  It's not valid JSON, fix the commas.  You probably mean "psqr" to be scoped?  Add an `@id` property and put the other properties under an embedded `@context` object.  Defining things as ".../v1/#property" might not be the best thing as far as future updates.
> > * The example in https://vpsqr.com/did-psqr/v1/ is not valid JSON, fix the commas.  It's also missing a context that defines the terms used.  The processed output semantic data is mostly empty.
> 
> Thanks for the review @davidlehn.
> 
> > fix the commas.
> > Done.
> 
> The JSON LD is only human readable at this point. Once it rises to the top of the development queue, we'll flesh it out and make it functional.
> 
> I did scope the `@context` under `psqr` and add `@id`, but a functional implementation of JSON LD is not yet present.

As this is not a blocker for this PR, we should probably take this elsewhere. I'm not sure where you'd like issues.  I think my comments were misunderstood.  The context at https://vpsqr.com/ns/did-psqr/v1/ is this:

```json
{
  "@context": {
    "psqr": {
      "publicIdentity": {
        "@id": "https://vpsqr.com/did-psqr/v1/#publicIdentity",
      },
      "publicKeys": {
        "@id": "https://w3c-ccg.github.io/security-vocab/#publicKeyJwk",
        "@container": "@set"
      },
      "permissions": {
        "@id": "https://vpsqr.com/did-psqr/v1/#permissions",
        "@container": "@set"
      },
      "updated": "https://vpsqr.com/did-psqr/v1/#updated"
    }
  }
}
```
But should probably be something like this with a [scoped context](https://www.w3.org/TR/json-ld/#scoped-contexts):
```json
{
  "@context": {
    "psqr": {
      "@id": "https://vpsqr.com/did-psqr/SOMETHINGSOMETHING",
      "@context": {
        "publicIdentity": {
          "@id": "https://vpsqr.com/did-psqr/v1/#publicIdentity"
        },
        "publicKeys": {
          "@id": "https://w3c-ccg.github.io/security-vocab/#publicKeyJwk",
          "@container": "@set"
        },
        "permissions": {
          "@id": "https://vpsqr.com/did-psqr/v1/#permissions",
          "@container": "@set"
        },
        "updated": "https://vpsqr.com/did-psqr/v1/#updated"
      }
    }
  }
}
```
And the example needs the context at the top level, so the "psqr" term gets expanded properly:
```json
{
    "@context": [
      "https://www.w3id.org/did/v1",
      "https://vpsqr.com/ns/did-psqr/v1"
    ],
    "id": "did:psqr:acme-publishing.com",
    "psqr": {
      "@id": "https://www.acme-publishing.com",
      ...
    }
}
```
And while I'm here... based on https://github.com/w3c-ccg/security-vocab/blob/main/contexts/security-v3-unstable.jsonld, I'm pretty sure you should be using https://w3id.org/security#publicKeyJwk instead of https://w3c-ccg.github.io/security-vocab/#publicKeyJwk.  I haven't kept up on what the latest state is of all that though.

I shall go off now and cry about the state of the tooling to make this all easier.

gribneau commented:
> As this is not a blocker for this PR, we should probably take this elsewhere. I'm not sure where you'd like issues.

We'll be going through license issues and exposing code as open source very soon, and we'll put it on github, at least as a mirror.

This will likely be revised during implementation (probably including removing the `v1`, but for the moment I've moved the scoping into the namespace and revised the JWK reference:
```
{
  "@context": {
    "psqr":{
      "@id": "https://vpsqr.com/did-psqr/v1/#psqr",
      "@context": {
        "publicIdentity": {
          "@id": "https://vpsqr.com/did-psqr/v1/#publicIdentity"
        },
        "publicKeys": {
          "@id": "https://www.w3.org/security/#publicKeyJwk",
          "@container": "@set"
        },
        "permissions": {
          "@id": "https://vpsqr.com/did-psqr/v1/#permissions",
          "@container": "@set"
        },
        "updated": "https://vpsqr.com/did-psqr/v1/#updated"
      }
    }
  }
}
```

OR13 commented:
`"@id": "https://vpsqr.com/did-psqr/v1/#publicIdentity"` probably want to remove the v1 from the permanent iri.

davidlehn commented:
> [...] and revised the JWK reference:
> 
> ```
>         "publicKeys": {
>           "@id": "https://www.w3.org/security/#publicKeyJwk",
>           "@container": "@set"
>         },
> ```

Should use `https://www.w3.org/security#publicKeyJwk` without the `/` since that's how it's defined and software may do direct string comparison on such values.

gribneau commented:
> `"@id": "https://vpsqr.com/did-psqr/v1/#publicIdentity"` probably want to remove the v1 from the permanent iri.

I added http 301 redirects to the most recent versions of the specifying documents and removed the version specifiers in the `@id` elements in the namespace JSON.


> Should use `https://www.w3.org/security#publicKeyJwk` without the `/`

I've removed the gratuitous trailing slash from all of the `@id` elements.

The PSQR `@id` elements should work properly as links, and no longer tie to a specific version.

OR13 commented:
https://www.w3.org/security#publicKeyJwk
https://w3id.org/security#publicKeyJwk

gribneau commented:
> https://www.w3.org/security#publicKeyJwk https://w3id.org/security#publicKeyJwk

I see. The `w3id.org` domain forwards. That makes sense.

@davidlehn mentioned a direct string matching issue. Is `w3id.org` acceptable in that case?

davidlehn commented:
Heh, oops, yeah, use w3id.org id. ;-)

gribneau commented:
> use w3id.org

Done.
